[EventPipe] End user_events session upon continuation_stream closure

mdh1418 opened the pull request:
User_events-based EventPipe sessions should end when the profiling process disconnects from the Diagnostic Port.

This PR does the following:
- Adds the poll_func to IpcStream vtables
- Adds `ds_ipc_stream_poll` based on `ds_ipc_poll`
- Pass the continuation stream to user_events sessions
- Extends the eventpipe streaming thread for user_events with `ep_ipc_continuation_stream_connection_closed` to detect disconnects.

### Testing

Manual E2E testing with the same C program as https://github.com/dotnet/runtime/pull/115265, except putting a `getchar` before closing the socket from the client side.

mdh1418 committed:
[EventPipe][DiagnosticServer] Add poll interface to IpcStreams

mdh1418 committed:
[EventPipe] Close user_events session upon continuation_stream closure

mdh1418 commented on src/native/eventpipe/ep-session.c:
```diff
@@ -100,28 +100,44 @@ EP_RT_DEFINE_THREAD_FUNC (streaming_thread)
-				break;
-			}
+		if (ep_session_type_uses_buffer_manager (session->session_type)) {
+			ep_rt_wait_event_handle_t *wait_event = ep_session_get_wait_event (session);
```
It's used only in the buffer-manager branch, in the check for `!events_written`.

mdh1418 commented on src/native/eventpipe/ds-ipc-pal-namedpipe.c:
```diff
@@ -834,12 +834,23 @@ ipc_stream_close_func (void *object)
+	void *object,
+	uint32_t timeout_ms)
+{
+	// Needs to be implemented.
+	return EP_IPC_POLL_EVENTS_UNKNOWN;
```
I haven't fully understood the `ds_ipc_poll` logic for named pipes, so wondering if we can just leave this unimplemented until somewhere actually needs to poll an IpcStream derivative on Windows.

mdh1418 commented on src/native/eventpipe/ds-eventpipe-protocol.c:
```diff
@@ -920,7 +920,7 @@ eventpipe_protocol_helper_collect_tracing (
 		payload->rundown_keyword,
 		payload->stackwalk_requested,
-		payload->session_type == EP_SESSION_TYPE_IPCSTREAM ? ds_ipc_stream_get_stream_ref (stream) : NULL,
+		ds_ipc_stream_get_stream_ref (stream),
```
This is reverting the line change from https://github.com/dotnet/runtime/pull/115265, and we know that stream cannot be NULL here due to `ep_return_false_if_nok (stream != NULL);` at the beginning of this function

noahfalk commented on src/native/eventpipe/ep-stream.c:
```diff
@@ -696,6 +708,21 @@ ep_stream_writer_write (
+ */
+
+bool
+ep_ipc_continuation_stream_connection_closed (IpcContinuationStream *ipc_continuation_stream)
```
```suggestion
ep_is_stream_connection_closed (IpcContinuationStream *ipc_continuation_stream)
```

mdh1418 replied:
Since we can just invoke the `ep_ipc_stream_poll_vcall` directly on the session's continuation stream, just made this a static function in `ep-session.c` and renamed to `session_is_stream_connection_closed`.

noahfalk commented on src/native/eventpipe/ep-stream.h:
```diff
@@ -355,5 +355,26 @@ ep_ipc_stream_writer_write (
 	uint32_t *bytes_written);
 
+/*
+ * IpcContinuationStream
```
Defining a new struct to wrap IpcStream seems unnecessary. Couldn't we just make the continuation_stream field in the session be type IpcStream*?

noahfalk replied:
I think IpcStream is meant to be a base class for different types of stream implementations, for example socket stream vs named pipe stream vs websocket stream. Its a bit less obvious because even though derived types have different implementations they are all reusing the same 'DiagnosticsIpcStream' type name. At some point in the future we could rename them SocketIpcStream, NamedPipeIpcStream, etc to make the inheritance pattern clearer. Sidestepping the naming, unless we are implementing a new transport I wouldn't expect a new stream derived type is needed.

noahfalk commented on src/native/eventpipe/ep-session.h:
```diff
@@ -71,6 +71,8 @@ struct _EventPipeSession_Internal {
 	volatile uint32_t ref_count;
 	// The user_events_data file descriptor to register Tracepoints and write user_events to.
 	int user_events_data_fd;
+	// The IPC continuation stream from initializing the session through the diagnostic server
```
```suggestion
	// The IPC continuation stream from initializing the session through the diagnostic server
	// Currently only initialized for user_events sessions.
```

mdh1418 replied:
Added

noahfalk commented on src/native/eventpipe/ep-session.c:
```diff
@@ -100,28 +100,44 @@ EP_RT_DEFINE_THREAD_FUNC (streaming_thread)
-			// Wait until it's time to sample again.
-			const uint32_t timeout_ns = 100000000; // 100 msec.
-			ep_rt_thread_sleep (timeout_ns);
+				// Wait until it's time to sample again.
```
```suggestion
				// Wait until it's time to poll again.
```
I see you copied it from above but 'sample' seems like an odd description of this :)

noahfalk commented on src/native/eventpipe/ds-ipc-pal-types.h:
```diff
@@ -22,14 +22,6 @@ typedef struct _DiagnosticsIpcStream DiagnosticsIpcStream;
  * Diagnostics IPC PAL Enums.
  */
 
-typedef enum {
```
Similar to the other comments, I'd suggest we leave this enum as-is and remove the EventPipe named version.

mdh1418 replied:
The ds-ipc-pal-types are not exposed to the EventPipe side, only `ds-ipc-pal-types.h` includes `ep-ipc-pal-types.h`. I moved the types to `ep-ipc-pal-types.h` to avoid a circular dependency, with the idea the `ep-ipc-pal-types` acts as the "common" pal types that bridges the DS side and EP side.

mdh1418 replied:
Ah, maybe its only caught by checked build? Or maybe my local had some caching even though I deleted the artifacts/bin/coreclr and artifacts/obj/coreclr folders
```
  /__w/1/s/src/native/eventpipe/ep-session.c:231:47: warning: nested extern declaration of ‘ep_ipc_stream_poll_vcall’ [-Wnested-externs]
  /__w/1/s/src/native/eventpipe/ep-session.c:231:81: error: ‘DS_IPC_TIMEOUT_INFINITE’ undeclared (first use in this function)
    231 |         DiagnosticsIpcPollEvents poll_event = ep_ipc_stream_poll_vcall (stream, DS_IPC_TIMEOUT_INFINITE);
        |                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~
  /__w/1/s/src/native/eventpipe/ep-session.c:231:81: note: each undeclared identifier is reported only once for each function it appears in
  /__w/1/s/src/native/eventpipe/ep-session.c:232:30: error: ‘DS_IPC_POLL_EVENTS_HANGUP’ undeclared (first use in this function)
    232 |         return poll_event == DS_IPC_POLL_EVENTS_HANGUP;
```

Going to move to `ep-ipc-pal-types.h`, keeping the same name.

lateralusX replied:
We shouldn't have a dependency from EP to DS, so if we have types that will be used by code in EP as well as DS, then it should go into the shared includes within EP "namespace".

noahfalk replied:
It sounds like we'd ideally have three layers here:
1. diagnostic server - handles incoming requests and dispatches them to the right component to get a response
2. event pipe
3. IPC - stream APIs that allow data to be read/written over different transports

Then the dependency tree (DAG really) might look like:
```
       DS
      /  \
     EP  |
       \ |
        IPC
```


Right now the IPC layer and diagnostic server layer are merged together so the dependencies from 1->2 and 2->3 form a cycle instead of a DAG. I liked @lateralusX's suggestion that DS_IPC_POLL_* could become IPC_POLL_* which moves in the direction of separating the IPC into its own layer rather than coupling it to DS.

mdh1418 commented on src/native/eventpipe/ds-ipc-pal-websocket.c:
```diff
@@ -425,12 +425,33 @@ ipc_stream_close_func (void *object)
+EventPipeIpcPollEvents
+ipc_stream_poll_func (
+	void *object,
+	uint32_t timeout_ms)
```
I was mainly trying to extract the polling mechanism from each pal's `ds_ipc_poll`, which in websockets scenario, the calls didn't seem to have a timeout option.

I had initially added this before deciding to forego the namedpipes implementation, but you're right. I think I'll also leave this one unimplemented until websockets really needs this function.

mdh1418 replied:
User_events is a Linux-only kernel feature that enables user space apps to emit telemetry in kernel space. EventPipe logic was extended in this [PR](https://github.com/dotnet/runtime/pull/115265) to allow user_events-based eventpipe sessions to be started.

I missed in that PR that the User_events eventpipe session should be closed when we detect that the client disconnects from the DiagnosticPort. Currently, that detection just checks whether events can be written to the IpcStream, but because User_events doesn't use the IpcStream to write events, we cannot leverage that logic. Instead, we are going to poll the current connection to check for a Hangup, and then stop the session if so.

To do so, I needed to add a poll_func on the IpcStream, so originally, I was attempting to implement the poll_func for all DiagnosticPort implementations, but now, I will only add it for sockets, and leave the NamedPipe + Websockets as unimplemented.

pavelsavara replied:
thanks

mdh1418 committed:
Address Feedback

Revert moving IpcPollEvents to ep-ipc-pal-types header
Use IpcStream base type directly in sessions
Omit websocket ipc_stream_poll_func implementation
Change continuation stream poll timeout to infinite
Change returned PollEvent to none upon timeout

mdh1418 commented on src/native/eventpipe/ep-session.c:
```diff
@@ -100,28 +100,44 @@ EP_RT_DEFINE_THREAD_FUNC (streaming_thread)
+				ep_rt_thread_sleep (timeout_ns);
 			}
+		} else if (session->session_type == EP_SESSION_TYPE_USEREVENTS) {
+			// User events session, write all user events tracepoints to the file.
```
Changed, thanks!

mdh1418 committed:
Move IpcPollEvents back to ep-ipc-pal-types header

mdh1418 committed:
Add relevant headers

ep-stream.h for ep_ipc_stream_writer_alloc that was transitively
included by ep-file.h
ep-ipc-stream.h for ep_ipc_stream_poll_vcall

lateralusX commented on src/native/eventpipe/ep-session.h:
```diff
@@ -71,6 +71,9 @@ struct _EventPipeSession_Internal {
 	int user_events_data_fd;
+	// The IPC continuation stream from initializing the session through the diagnostic server
+	// Currently only initialized for user_events sessions.
+	IpcStream *stream;
```
We don't need to do it in this PR, but going forward I believe we should do different versions of this struct depending on the type and move specific fields into sub types instead of having a fields in one struct where type dictates what fields is applicable without any further information.

mdh1418 replied:
Right, I think we can have a Refactoring PR where we split up the versions of the struct, also migrate the IPC types to its own file, and move the user_events_data fd into the Stream structure

mdh1418 replied:
So [IPC namespace](https://github.com/dotnet/runtime/pull/117435/#discussion_r2199927716), [Clarify the IpcStream inheritance naming](https://github.com/dotnet/runtime/pull/117435/#discussion_r2202296190), this, and the followup work from the main user_events PR (https://github.com/dotnet/runtime/pull/115265#discussion_r2139467105 ,https://github.com/dotnet/runtime/pull/115265#discussion_r2123840111, https://github.com/dotnet/runtime/pull/115265#discussion_r2123349055)

lateralusX commented on src/native/eventpipe/ep-session.c:
```diff
@@ -183,25 +205,35 @@ static
+{
+	EP_ASSERT (stream != NULL);
+	DiagnosticsIpcPollEvents poll_event = ep_ipc_stream_poll_vcall (stream, DS_IPC_TIMEOUT_INFINITE);
+	return poll_event == DS_IPC_POLL_EVENTS_HANGUP;
```
Should we handle errors as close as well?

mdh1418 replied:
I wasn't sure previously if there was a chance there could be an error from a poll operation even though the connection didn't close where a following poll retry would succeed. I think we can default to just treating both Error and hangup as connection closed.

mdh1418 committed:
Address feedback

Remove prefix from IpcPollEvents enums and infinite timeout define
Expand connection closed detection to error case

lateralusX reviewed (approved):
LGTM!
